### PR TITLE
Feat(button): add auto-loading prop

### PR DESCRIPTION
**Pull request by doom-9**
## Checklist

---

- [ ] Fix linting errors
- [ ] Tests have been added / updated (or snapshots)

## Change information

---

### CODE REVIEW

**Comment by doom-9** [comment]
close https://github.com/varletjs/varlet/issues/281

**Comment by doom-9** [comment]
test('test autoLoading', async () => {
  const callBack = () => {
    return new Promise((resolve) => {
      setTimeout(() => {
        resolve()
      }, 2000)
    })
  }
  const wrapper = mount(VarButton, {
    props: {
      onClick: callBack,
      onTouchstart: callBack,
      autoLoading: true,
    },
  })

  await wrapper.trigger('click')
  expect(wrapper.find('.var-loading').exists()).toBeTruthy()

  await delay(2500)
  expect(wrapper.find('.var-loading').exists()).toBeFalsy()

  await wrapper.trigger('touchstart')
  expect(wrapper.find('.var-loading').exists()).toBeTruthy()

  await delay(2500)
  expect(wrapper.find('.var-loading').exists()).toBeFalsy()

  wrapper.unmount()
})

**Comment by doom-9** [comment]
CleanShot 2022-01-17 at 22 25 13@2x

File: packages/varlet-ui/src/button/Button.vue [hunk 1]
```diff
@@ -48,29 +48,44 @@ export default defineComponent({
   directives: { Ripple },
   props,
   setup(props) {
+    let loadingRef = props.autoLoading ? ref(false) : toRef(props, 'loading')
```
Comment (haoziqaq) [inline packages/varlet-ui/src/button/Button.vue:51]: Component behavior needs to be considered when props.autoLoading changes

File: packages/varlet-ui/src/button/Button.vue [hunk 2]
```diff
@@ -48,29 +48,44 @@ export default defineComponent({
...
+      const { disabled, onClick, autoLoading } = props
 
-      if (loading || disabled) {
+      if (loadingRef.value || disabled) {
         return
       }
 
-      onClick?.(e)
+      const returnValue = onClick?.(e)
+      if (autoLoading && returnValue instanceof Promise) {
```
Comment (haoziqaq) [inline packages/varlet-ui/src/button/Button.vue:61]: The behavior of handling autoloading needs to be extracted into a function

Comment (haoziqaq) [reply packages/varlet-ui/src/button/Button.vue:61]: returnValue instanceof Promise should be extracted into a function called isPromise and put in utils/shared

Comment (doom-9) [reply packages/varlet-ui/src/button/Button.vue:61]: 👌

**Comment by haoziqaq** [comment]
Thank you, I merged it first, and someone will optimize this part of the code later.

**Pull request merged**